### PR TITLE
Add Dynamic command support to AdvancedLogDumper Application

### PR DESCRIPTION
**Pull request by sureshkumarpMSFT**
## Description

- Added support to invoke AdvlogDump as in-built UEFI Shell application using EFI_SHELL_DYNAMIC_COMMAND_PROTOCOL

- Created new INF file to add this support

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [x ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

- Invoked "Advlogdump" from UEFI Shell and verified LogDumper functionality 
## Integration

Add the following to DSC file
  AdvLoggerPkg/Application/AdvancedLogDumper/AdvancedLogDumperDynamicCommand.inf {
   <PcdsFixedAtBuild>
      gEfiShellPkgTokenSpaceGuid.PcdShellLibAutoInitialize|FALSE
  } 

in FDF file

  INF AdvLoggerPkg/Application/AdvancedLogDumper/AdvancedLogDumperDynamicCommand.inf

### CODE REVIEW

**Comment by sureshkumarpMSFT** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: AdvLoggerPkg/Application/AdvancedLogDumper/AdvancedLogDumperDynamicCommand.inf [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  UefiBootServicesTableLib
+  UefiDriverEntryPoint
+  HiiLib
+
+[Protocols]
+  gEfiHiiPackageListProtocolGuid                  ## CONSUMES
+  gEfiShellDynamicCommandProtocolGuid             ## PRODUCES
+
+[Depex]
+  TRUE
```
Comment (os-d) [inline AdvLoggerPkg/Application/AdvancedLogDumper/AdvancedLogDumperDynamicCommand.inf]: nit: add an empty line to EOF

File: AdvLoggerPkg/Application/AdvancedLogDumper/LogDumper.uni [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/** @file
+AdvancedLogDumper.uni
+
+String definitions for AdvancedLogDumper command
+
+Copyright (C) Microsoft Corporation. All rights reserved.
```
Comment (os-d) [inline AdvLoggerPkg/Application/AdvancedLogDumper/LogDumper.uni]:
```suggestion
Copyright (C) Microsoft Corporation.
```

File: AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperCommon.c [hunk 3]
```diff
@@ -0,0 +1,341 @@
+/** @file
+LogDumper.c
+
+This application will dump the AdvancedLog to a file.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
```
Comment (os-d) [inline AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperCommon.c]:
```suggestion
Copyright (C) Microsoft Corporation.
```

File: AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperDynamiCommand.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
+/** @file
+LogDumper.c
+
+This application will dump the AdvancedLog to a file.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
```
Comment (os-d) [inline AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperDynamiCommand.c]:
```suggestion
Copyright (C) Microsoft Corporation.
```

File: AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperDynamiCommand.c [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                  &gEfiShellDynamicCommandProtocolGuid,
+                  &mAdvLogDumperDynamicCommand
+                  );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  HiiRemovePackages (gAdvLogHiiHandle);
+  return EFI_SUCCESS;
+}
```
Comment (os-d) [inline AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperDynamiCommand.c]: nit: add empty line at EOF

**Review by os-d — APPROVED** [review]

File: AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperCommon.c [hunk 6]
```diff
@@ -0,0 +1,341 @@
+/** @file
+LogDumper.c
```
Comment (kuqin12) [inline AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperCommon.c]:
```suggestion
LogDumperCommon.c
```

**Review by kuqin12 — APPROVED** [review]

File: AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperDynamiCommand.c
Comment (makubacki) [inline AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperDynamiCommand.c:1]: Can you fix the typo in this file name (`LogDumperDynamiCommand.c` to `LogDumperDynamicCommand.c`).

File: AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperDynamiCommand.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
+/** @file
+LogDumper.c
```
Comment (makubacki) [inline AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperDynamiCommand.c]: File name is unnecessary here, but if present, it should match the actual file name.

File: AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperCommon.c [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+/**
+ * Requests the Advanced Logger variables and dumps the raw file to a file
+ * @param[in] FileHandle         The handle of the file we want to write to
+ * @param[in] Verbose            Whether debugging statements should be written out to console
+ * @retval EFI_SUCCESS           We were able to write to the file
+ * @retval EFI_OUT_OF_RESOURCES  We failed to allocate a buffer
+ * @retval EFI_NOT_FOUND         If AdvLogger didn't create a buffer or isn't installed
+ * @retval EFI_INVALID_PARAMETER The FileHandle was bad or null
+ * @retval Others                Errors passed from GetVariable or ShellWriteFile
+ */
```
Comment (makubacki) [inline AdvLoggerPkg/Application/AdvancedLogDumper/LogDumperCommon.c]: To match current conventions.

```suggestion
/**
  Requests the Advanced Logger variables and dumps the raw file to a file

  @param[in] FileHandle         The handle of the file we want to write to
  @param[in] Verbose            Whether debugging statements should be written out to console

  @retval EFI_SUCCESS           We were able to write to the file
  @retval EFI_OUT_OF_RESOURCES  We failed to allocate a buffer
  @retval EFI_NOT_FOUND         If AdvLogger didn't create a buffer or isn't installed
  @retval EFI_INVALID_PARAMETER The FileHandle was bad or null
  @retval Others                Errors passed from GetVariable or ShellWriteFile
 */

```

**Review by makubacki — APPROVED** [review]

**Comment by sureshkumarpMSFT** [comment]
@os-d @makubacki @kuqin12 

Thanks for the review and comments.  

-   Since i have added a DXE_DRIVER and using APPLICATION lib the Pipeline build reported issues (due to the changes in AdvLoggerPkg.dsc) and I had to do multiple commits to fix each one of them. Sorry for the inconvenience

-  Now Pipeline builds are OK

Please let me know the next steps on this

**Pull request merged**